File: frontend/src/lib/chat-attachment-policy.ts
/**
 * Chat attachment mode policy (D-082 / D-083).
 * Keep extension lists aligned with backend `document_parsers.py`.
 */

export const LOCAL_TEXT_EXTENSIONS = new Set([
  ".txt",
  ".md",
  ".markdown",
  ".html",
  ".htm",
  ".csv",
  ".log",
  ".ini",
  ".cfg",
  ".conf",
  ".toml",
  ".yaml",
  ".yml",
  ".json",
  ".xml",
  ".css",
  ".scss",
  ".less",
  ".py",
  ".pyi",
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".mjs",
  ".cjs",
  ".java",
  ".go",
  ".rs",
  ".c",
  ".cc",
  ".cpp",
  ".h",
  ".hpp",
  ".cs",
  ".rb",
  ".php",
  ".swift",
  ".kt",
  ".kts",
  ".scala",
  ".sql",
  ".sh",
  ".bash",
  ".zsh",
  ".ps1",
  ".r",
  ".vue",
  ".svelte",
  ".dart",
  ".lua",
  ".pl",
  ".pm",
]);

export const WHITELIST_DOCUMENT_EXTENSIONS = new Set([
  ".pdf",
  ".doc",
  ".docx",
  ".pptx",
  ".xls",
  ".xlsx",
]);

export const IMAGE_EXTENSIONS = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".webp",
  ".gif",
  ".tif",
  ".tiff",
]);
export const AUDIO_EXTENSIONS = new Set([
  ".mp3",
  ".m4a",
  ".wav",
  ".ogg",
  ".flac",
  ".aac",
]);
export const VIDEO_EXTENSIONS = new Set([
  ".mp4",
  ".mov",
  ".avi",
  ".webm",
  ".mkv",
  ".flv",
  ".wmv",
]);
export const SPECIAL_BINARY_EXTENSIONS = new Set([
  ".exe",
  ".dll",
  ".so",
  ".dylib",
  ".msi",
  ".bat",
  ".cmd",
  ".com",
  ".scr",
  ".sys",
  ".apk",
  ".dmg",
  ".iso",
  ".img",
  ".bin",
  ".appimage",
]);

export function fileExtension(name: string): string {
  const base = name.split(/[\\/]/).pop() ?? name;
  const index = base.lastIndexOf(".");
  if (index <= 0) return "";
  return base.slice(index).toLowerCase();
}

export function isImageNameOrMime(name: string, mime?: string | null): boolean {
  const type = (mime ?? "").toLowerCase().split(";", 1)[0].trim();
  return type.startsWith("image/") || IMAGE_EXTENSIONS.has(fileExtension(name));
}

export function isAudioNameOrMime(name: string, mime?: string | null): boolean {
  const type = (mime ?? "").toLowerCase().split(";", 1)[0].trim();
  return type.startsWith("audio/") || AUDIO_EXTENSIONS.has(fileExtension(name));
}

export function isVideoNameOrMime(name: string, mime?: string | null): boolean {
  const type = (mime ?? "").toLowerCase().split(";", 1)[0].trim();
  return type.startsWith("video/") || VIDEO_EXTENSIONS.has(fileExtension(name));
}

export function isSpecialBinaryName(name: string): boolean {
  return SPECIAL_BINARY_EXTENSIONS.has(fileExtension(name));
}

export function isFastThinkingWhitelistDocument(name: string): boolean {
  if (isSpecialBinaryName(name)) return false;
  const extension = fileExtension(name);
  return (
    LOCAL_TEXT_EXTENSIONS.has(extension) ||
    WHITELIST_DOCUMENT_EXTENSIONS.has(extension)
  );
}

export type AttachmentBlockReason =
  | "unsupported"
  | "audio_needs_asr"
  | "document_not_ready"
  | null;

export function classifyNonAgentAttachment(options: {
  name: string;
  mime?: string | null;
  parseStatus?: string | null;
  asrAvailable: boolean;
  requireReady?: boolean;
}): { ok: boolean; reason: AttachmentBlockReason } {
  const {
    name,
    mime,
    parseStatus,
    asrAvailable,
    requireReady = false,
  } = options;
  if (isSpecialBinaryName(name)) return { ok: false, reason: "unsupported" };
  if (isImageNameOrMime(name, mime) || isVideoNameOrMime(name, mime)) {
    return { ok: true, reason: null };
  }
  if (isAudioNameOrMime(name, mime)) {
    return asrAvailable
      ? { ok: true, reason: null }
      : { ok: false, reason: "audio_needs_asr" };
  }
  if (!isFastThinkingWhitelistDocument(name)) {
    return { ok: false, reason: "unsupported" };
  }
  if (requireReady && parseStatus !== "indexed") {
    return { ok: false, reason: "document_not_ready" };
  }
  return { ok: true, reason: null };
}

export function nonAgentAttachmentBlockedMessage(names: string[]): string {
  const uniqueNames = [...new Set(names.filter(Boolean))];
  const preview = uniqueNames.slice(0, 5).map((name) => `「${name}」`).join("、");
  const remainder = uniqueNames.length > 5 ? ` 等 ${uniqueNames.length} 个文件` : "";
  return `极速/思考模式不支持以下附件：${preview}${remainder}。请切换到智能体模式，或者删除不受支持的文件后再发送。`;
}

File: backend/app/services/chat_attachment_policy.py
"""Chat attachment mode policy (D-082 / D-083).

Non-agent (fast/thinking) turns may use files that have a safe text extraction
path, direct image input, video input the selected model can actually consume
(natively or through a video-capable vision companion), or audio when
stored-file ASR is configured. Anything else is still accepted as an attachment
and is rejected at send time with a typed error that points at Agent mode, which
may attach any stored workspace file. Executables, scripts and disk images are
the single exception: they are refused in every mode.
"""

from __future__ import annotations

from pathlib import Path

from app.domain.models import FileRecord
from app.services.document_parsers import (
    BUILT_IN_DOCUMENT_EXTENSIONS,
    ISOLATED_DOCUMENT_EXTENSIONS,
    LOCAL_TEXT_EXTENSIONS,
)


WHITELIST_DOCUMENT_EXTENSIONS = (
    BUILT_IN_DOCUMENT_EXTENSIONS | ISOLATED_DOCUMENT_EXTENSIONS
)
OPTIONAL_IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".webp", ".gif", ".tif", ".tiff"}
AUDIO_EXTENSIONS = {".mp3", ".m4a", ".wav", ".ogg", ".flac", ".aac"}
VIDEO_EXTENSIONS = {".mp4", ".mov", ".avi", ".webm", ".mkv", ".flv", ".wmv"}

# Explicitly never allowed in fast/thinking (even if metadata is malformed).
SPECIAL_BINARY_EXTENSIONS = {
    ".exe",
    ".dll",
    ".so",
    ".dylib",
    ".msi",
    ".bat",
    ".cmd",
    ".com",
    ".scr",
    ".sys",
    ".apk",
    ".dmg",
    ".iso",
    ".img",
    ".bin",
    ".appimage",
}


def file_extension(name: str) -> str:
    return Path(name or "").suffix.casefold()


def is_image_attachment(file: FileRecord) -> bool:
    mime = (file.mime_type or "").casefold().split(";", 1)[0].strip()
    if mime.startswith("image/"):
        return True
    return file_extension(file.original_name) in OPTIONAL_IMAGE_EXTENSIONS


def is_audio_attachment(file: FileRecord) -> bool:
    mime = (file.mime_type or "").casefold().split(";", 1)[0].strip()
    if mime.startswith("audio/"):
        return True
    return file_extension(file.original_name) in AUDIO_EXTENSIONS


def is_video_attachment(file: FileRecord) -> bool:
    mime = (file.mime_type or "").casefold().split(";", 1)[0].strip()
    if mime.startswith("video/"):
        return True
    return file_extension(file.original_name) in VIDEO_EXTENSIONS


def is_special_binary_attachment(file: FileRecord) -> bool:
    return file_extension(file.original_name) in SPECIAL_BINARY_EXTENSIONS


def is_fast_thinking_whitelist_document(file: FileRecord) -> bool:
    """True when the file has a supported safe text extraction path."""

    if is_image_attachment(file) or is_audio_attachment(file) or is_video_attachment(file):
        return False
    if is_special_binary_attachment(file):
        return False
    extension = file_extension(file.original_name)
    return extension in LOCAL_TEXT_EXTENSIONS or extension in WHITELIST_DOCUMENT_EXTENSIONS


def is_inline_text_attachment(file: FileRecord) -> bool:
    """True when the attachment travels to the model as its complete stored text.

    Text, code and markup attachments (including HTML/Markdown artifacts) are
    inlined by ``ChatService._inline_text_attachment_context`` instead of being
    retrieved through the sparse index, so they never need a text index and must
    not be gated behind one.
    """

    if is_special_binary_attachment(file):
        return False
    if is_image_attachment(file) or is_audio_attachment(file) or is_video_attachment(file):
        return False
    return file_extension(file.original_name) in LOCAL_TEXT_EXTENSIONS


def classify_non_agent_attachment(
    file: FileRecord,
    *,
    asr_available: bool,
) -> str:
    """Return a stable readiness class for fast/thinking attachments."""

    if is_special_binary_attachment(file):
        return "unsupported"
    if is_image_attachment(file):
        return "image"
    if is_video_attachment(file):
        return "video"
    if is_audio_attachment(file):
        return "audio_ok" if asr_available else "audio_needs_asr"
    if not is_fast_thinking_whitelist_document(file):
        return "unsupported"
    if file.parse_status == "indexed":
        return "document_ready"
    if is_inline_text_attachment(file):
        # Text/code attachments (including HTML/Markdown artifacts) are delivered
        # to the model as their complete stored text, so they never need a text
        # index. Requiring one here blocked exactly the attachments that are
        # meant to be used as-is.
        return "text_inline"
    return "document_not_ready"


def non_agent_attachment_error(
    file: FileRecord,
    classification: str,
) -> tuple[str, str]:
    """Return (error_code, message) for a blocked non-agent attachment."""

    name = file.original_name or file.id
    suffix = "请切换到智能体模式，或删除不受支持的文件后再用极速/思考发送。"
    if classification == "audio_needs_asr":
        return (
            "transcription_provider_unavailable",
            f"音频「{name}」无法使用已配置的文件 ASR 转写。{suffix}",
        )
    if classification == "document_not_ready":
        return (
            "attachment_not_indexed",
            f"文件「{name}」尚未完成文本解析/索引。{suffix}",
        )
    return (
        "attachment_mode_unsupported",
        f"文件「{name}」不是极速/思考模式支持的附件类型。{suffix}",
    )
